src/tallies/filter_weight.cpp

```diff
@@ -0,0 +1,61 @@
+#include "openmc/tallies/filter_weight.h"
+
+#include <fmt/core.h>
+
+#include "openmc/capi.h"
+#include "openmc/search.h"
+#include "openmc/constants.h"
+#include "openmc/settings.h"
+#include "openmc/xml_interface.h"
+
+namespace openmc {
+
+//==============================================================================
+// WeightFilter implementation
+//==============================================================================
+
+void WeightFilter::from_xml(pugi::xml_node node)
+{
+  auto bins = get_node_array<double>(node, "bins");
+  this->set_bins(bins);
+}
+
+void WeightFilter::set_bins(gsl::span<const double> bins)
+{
+   // Clear existing bins
+   bins_.clear();
+   bins_.reserve(bins.size());
+ 
+   // Copy bins
+   for (gsl::index i = 0; i < bins.size(); ++i) {
+     bins_.push_back(bins[i]);
+   }
+ 
+   n_bins_ = bins_.size() -1;

```

Comment:
As an extra layer of protection, I'd suggest `assert(std::is_sorted( ...` here. By the way, you can remove `gsl::` from this since it was removed in #3225 .